cpp/arcticdb/version/version_map_batch_methods.hpp

```diff
@@ -173,28 +173,94 @@ struct MapRandomAccessWrapper {
     }
 };
 
+template <typename Inputs, typename TaskSubmitter>
+inline void submit_cpu_task_for_range(const Inputs& inputs, TaskSubmitter submitter) {
+    const auto window_size = async::TaskScheduler::instance()->cpu_thread_count();
+
+    auto futures = folly::window(inputs, [&submitter](const auto &input) {
+        return submitter(input);
+    }, window_size);
+
+    auto collected_futs = folly::collectAll(futures).get();
+    std::optional<std::string> all_exceptions;
+    for (auto&& collected_fut: collected_futs) {
+        if (!collected_fut.hasValue()) {
+            all_exceptions = all_exceptions.value_or("").append(collected_fut.exception().what().toStdString()).append("\n");
+        }
+    }
+    internal::check<ErrorCode::E_RUNTIME_ERROR>(!all_exceptions.has_value(), all_exceptions.value_or(""));
+}
+
+template <typename Task, typename... T>
+struct IterationTask : async::BaseTask {
+    Task func_;
+    std::tuple<T...> items;
+    explicit IterationTask(
+            Task&& func,
+            T&&... t) :
+            func_(std::move(func)),
+            items{std::make_tuple(std::forward<T>(t)...)} {
+    }
+
+    ARCTICDB_MOVE_ONLY_DEFAULT(IterationTask)
+
+    void operator()() {
+        std::apply(func_, std::forward<decltype(items)>(items));
+    }
+};
+
 // This version assumes that there is only one version per symbol, so no need for the state machine below
 inline std::shared_ptr<std::unordered_map<StreamId, AtomKey>> batch_get_specific_version(
     const std::shared_ptr<Store>& store,
     const std::shared_ptr<VersionMap>& version_map,
     const std::map<StreamId, VersionId>& sym_versions,
-    bool include_deleted = true) {
+    bool include_deleted = true,
+    bool include_referenced_in_snapshot_only = false) {
     ARCTICDB_SAMPLE(BatchGetLatestVersion, 0)
     auto output = std::make_shared<std::unordered_map<StreamId, AtomKey>>();
-    auto mutex = std::make_shared<std::mutex>();
+    auto output_mutex = std::make_shared<std::mutex>();
+    auto tombstoned_vers = std::make_shared<std::vector<std::pair<StreamId, AtomKey>>>(); // symbol, version key
```

Comment:
This comment is inaccurate, the second entry of the pair is an index key.
I think this structure would also be clearer as an unordered map than a vector of pairs.

Reply:
Fair enough